File: telegram_bot/engineerbot/bot/bot.go
package bot

import (
	"bytes"
	"encoding/json"
	"log"
	"net/http"
	"strings"
	"time"

	tgbotapi "github.com/go-telegram-bot-api/telegram-bot-api"
)

type Bot struct {
	*tgbotapi.BotAPI
}

type EndData struct {
	Timestamp string  `json:"completion_date"`
	Comment   string `json:"comment,omitempty"`
}

type StartData struct {
	Timestamp string `json:"start_date"`
}

func NewBot(token string) (*Bot, error) {
	tgBot, err := tgbotapi.NewBotAPI(token)
	if err != nil {
		return nil, err
	}

	bot := &Bot{tgBot}

	go bot.startUpdatesLoop()

	return bot, nil
}

func (b *Bot) startUpdatesLoop() {
	u := tgbotapi.NewUpdate(0)
	u.Timeout = 60

	updates, err := b.BotAPI.GetUpdatesChan(u)
	if err != nil {
		log.Panic(err)
	}

	for update := range updates {
		// Handle updates here
		if update.CallbackQuery != nil {
			callbackData := update.CallbackQuery.Data
			callbackParts := strings.Split(callbackData, ":")
			action := callbackParts[0]
			ticketID := callbackParts[1]

			// Call your corresponding methods for each action
			switch action {
			case "TookTicket":
				actionData := StartData{
					Timestamp: time.Now().UTC().Format("2006-01-02T15:04:05.999Z07:00"),
				}
				jsonData, err := json.Marshal(actionData)
				if err != nil {
					log.Panic(err)
				}
				log.Println("Starting sending request")
				req, err := http.NewRequest("PUT", "http://localhost:3030/update_start_date/"+ticketID, bytes.NewBuffer(jsonData))
				if err != nil {
					log.Panic(err)
				}
				req.Header.Set("Content-Type", "application/json")

				client := &http.Client{}
				resp, err := client.Do(req)
				if err != nil {
					log.Panic(err)
				}
				if resp.StatusCode == http.StatusOK{
					msg := tgbotapi.NewMessage(update.CallbackQuery.Message.Chat.ID, "You took the ticket!")
					if _, err := b.BotAPI.Send(msg); err != nil {
						log.Panic(err)
					}
					break
				}
				resp.Body.Close()

			case "Finished":
				msg := tgbotapi.NewMessage(update.CallbackQuery.Message.Chat.ID, "Would you like to add a comment?")
				msg.ReplyMarkup = tgbotapi.NewInlineKeyboardMarkup(
					tgbotapi.NewInlineKeyboardRow(
						tgbotapi.NewInlineKeyboardButtonData("Add comment", "AddComment:"+ticketID),
						tgbotapi.NewInlineKeyboardButtonData("Skip comment", "SkipComment:"+ticketID),
					),
				)

				if _, err := b.BotAPI.Send(msg); err != nil {
					log.Panic(err)
				}
				

			case "AddComment":
				// do nothing, just wait for the next message

			case "SkipComment":
				actionData := EndData{
					Timestamp: time.Now().UTC().Format("2006-01-02T15:04:05.999Z07:00"),
				}
				jsonData, err := json.Marshal(actionData)
				if err != nil {
					log.Panic(err)
				}
				log.Println("Starting sending request")
				req, err := http.NewRequest("PUT", "http://localhost:3030/update_completion_date/"+ticketID, bytes.NewBuffer(jsonData))
				if err != nil {
					log.Panic(err)
				}
				req.Header.Set("Content-Type", "application/json")

				client := &http.Client{}
				resp, err := client.Do(req)
				if err != nil {
					log.Panic(err)
				}
				if resp.StatusCode == http.StatusOK{
					msg := tgbotapi.NewMessage(update.CallbackQuery.Message.Chat.ID, "You finished the ticket!")
					if _, err := b.BotAPI.Send(msg); err != nil {
						log.Panic(err)
					}
					break
				} else {
					msg := tgbotapi.NewMessage(update.CallbackQuery.Message.Chat.ID, "Problem with update request")
					if _, err := b.BotAPI.Send(msg); err != nil {
						log.Panic(err)
					}
				}
				resp.Body.Close()

			}

			continue
		}

		if update.Message != nil && update.Message.IsCommand() {
			// Handle comments
			ticketID := update.Message.Command()
			//comment := update.Message.CommandArguments()

			actionData := EndData{
				Timestamp: time.Now().UTC().Format("2006-01-02T15:04:05.999Z07:00"),
			}

			jsonData, err := json.Marshal(actionData)
			if err != nil {
				log.Panic(err)
			}

			log.Println("Starting sending request")
				req, err := http.NewRequest("PUT", "http://localhost:3030/update_completion_date/"+ticketID, bytes.NewBuffer(jsonData))
				if err != nil {
					log.Panic(err)
				}
				req.Header.Set("Content-Type", "application/json")

				client := &http.Client{}
				resp, err := client.Do(req)
				if err != nil {
					log.Panic(err)
				}
				if resp.StatusCode == http.StatusOK{
					msg := tgbotapi.NewMessage(update.CallbackQuery.Message.Chat.ID, "You finished the ticket!")
					if _, err := b.BotAPI.Send(msg); err != nil {
						log.Panic(err)
					}
					break
				} else {
					msg := tgbotapi.NewMessage(update.CallbackQuery.Message.Chat.ID, "Problem with update request")
					if _, err := b.BotAPI.Send(msg); err != nil {
						log.Panic(err)
					}
				}
				resp.Body.Close()
			continue
		}
	}
}

File: telegram_bot/engineerbotv2/bot.py
import json
import logging
import requests
from datetime import datetime
from telegram import Update, InlineKeyboardMarkup, InlineKeyboardButton, Bot
from telegram.ext import Updater, CallbackContext, CommandHandler, CallbackQueryHandler, MessageHandler, Filters

class Bot:

    def __init__(self, token):
        self.updater = Updater(token=token, use_context=True)
        self.bot = Bot(token)

    def start_updates_loop(self):
        dp = self.updater.dispatcher
        dp.add_handler(CallbackQueryHandler(self.handle_callback))
        dp.add_handler(MessageHandler(Filters.command, self.handle_command))
        self.updater.start_polling()

    def handle_callback(self, update: Update, context: CallbackContext):
        query = update.callback_query
        query.answer()

        data_parts = query.data.split(":")
        action = data_parts[0]
        ticket_id = data_parts[1]

        if action == 'TookTicket':
            start_data = {
                'Timestamp': datetime.now().isoformat(),
            }
            self.update_data('http://localhost:3030/update_start_date/'+ticket_id, start_data)
            query.edit_message_text(text="You took the ticket!")

        elif action == 'Finished':
            keyboard = [
                [
                    InlineKeyboardButton("Add comment", callback_data='AddComment:'+ticket_id),
                    InlineKeyboardButton("Skip comment", callback_data='SkipComment:'+ticket_id),
                ]
            ]
            reply_markup = InlineKeyboardMarkup(keyboard)
            context.bot.send_message(chat_id=query.message.chat_id, text="Would you like to add a comment?", reply_markup=reply_markup)

        elif action == 'SkipComment':
            end_data = {
                'Timestamp': datetime.now().isoformat(),
            }
            self.update_data('http://localhost:3030/update_completion_date/'+ticket_id, end_data)
            query.edit_message_text(text="You finished the ticket!")

    def handle_command(self, update: Update, context: CallbackContext):
        ticket_id = update.message.text[1:]  # remove leading '/'
        end_data = {
            'Timestamp': datetime.now().isoformat(),
        }
        self.update_data('http://localhost:3030/update_completion_date/'+ticket_id, end_data)
        context.bot.send_message(chat_id=update.message.chat_id, text="You finished the ticket!")

    def update_data(self, url, data):
        headers = {'content-type': 'application/json'}
        response = requests.put(url, data=json.dumps(data), headers=headers)
        if response.status_code != 200:
            logging.error("Failed to update data.")

    def send_message(self, chat_id, text):
        self.bot.send_message(chat_id=chat_id, text=text)
